src/XMonad/Core.hs

```diff
@@ -552,19 +567,149 @@ getXMonadDataDir :: X String
+libFileName Directories{ cfgDir } = cfgDir </> "lib"
+
+buildScriptFileName, stackYamlFileName :: Directories -> FilePath
+buildScriptFileName Directories{ cfgDir } = cfgDir </> "build"
+stackYamlFileName   Directories{ cfgDir } = cfgDir </> "stack.yaml"
+
+-- | Compilation method for xmonad configuration.
+data Compile = CompileGhc | CompileStackGhc FilePath | CompileScript FilePath
+    deriving (Show)
+
+-- | Detect compilation method by looking for known file names in xmonad
+-- configuration directory.
+detectCompile :: Directories -> IO Compile
+detectCompile dirs = tryScript <|> tryStack <|> useGhc
+  where
+    buildScript = buildScriptFileName dirs
+    stackYaml = stackYamlFileName dirs
+
+    tryScript = do
+        guard =<< doesFileExist buildScript
+        isExe <- isExecutable buildScript
+        if isExe
+          then do
+            trace $ "XMonad will use build script at " <> show buildScript <> " to recompile."
+            pure $ CompileScript buildScript
+          else do
+            trace $ "XMonad will not use build script, because " <> show buildScript <> " is not executable."
+            trace $ "Suggested resolution to use it: chmod u+x " <> show buildScript
+            empty
+
+    tryStack = do
+        guard =<< doesFileExist stackYaml
+        canonStackYaml <- canonicalizePath stackYaml
+        trace $ "XMonad will use stack ghc --stack-yaml " <> show canonStackYaml <> " to recompile."
+        pure $ CompileStackGhc canonStackYaml
+
+    useGhc = do
+        trace $ "XMonad will use ghc to recompile, because neither "
+                <> show buildScript <> " nor " <> show stackYaml <> " exists."
+        pure CompileGhc
+
+    isExecutable f = E.catch (executable <$> getPermissions f) (\(SomeException _) -> return False)
+
+-- | Should we recompile xmonad configuration? Is it newer than the compiled
+-- binary?
+shouldCompile :: Directories -> Compile -> IO Bool
+shouldCompile dirs CompileGhc = do
+    libTs <- mapM getModTime . Prelude.filter isSource =<< allFiles (libFileName dirs)
+    srcT <- getModTime (srcFileName dirs)
+    binT <- getModTime (binFileName dirs)
+    if any (binT <) (srcT : libTs)
+        then True <$ trace "XMonad recompiling because some files have changed."
+        else False <$ trace "XMonad skipping recompile because it is not forced (e.g. via --recompile), and neither xmonad.hs nor any *.hs / *.lhs / *.hsc files in lib/ have been changed."
+  where
+    isSource = flip elem [".hs",".lhs",".hsc"] . takeExtension
+    allFiles t = do
+        let prep = map (t</>) . Prelude.filter (`notElem` [".",".."])
+        cs <- prep <$> E.catch (getDirectoryContents t) (\(SomeException _) -> return [])
+        ds <- filterM doesDirectoryExist cs
+        concat . ((cs \\ ds):) <$> mapM allFiles ds
+shouldCompile dirs CompileStackGhc{} = do
+    stackYamlT <- getModTime (stackYamlFileName dirs)
+    binT <- getModTime (binFileName dirs)
+    if binT < stackYamlT
```

Comment:
I wonder if we should perhaps also execute `stack build` here?  If something in the stack yaml changes, it's likely that this is either adding a new dependency, or bumping the resolver.  In this case, we should rebuild everything before compiling the config, as we would get errors otherwise (because e.g. `xmonad-contrib` is just not built for the new resolver) and would have to manually invoke `stack build` (or `stack install` or something similar).  This might not be so obvious behaviour to users

Reply:
Oh, indeed! I don't know why I thought `stack ghc` would do this automatically. Probably just tired. :-)

Reply:
3f6569db66f3a2ac51005258a8ba1544d5ae1a6a should do it

Reply:
I'm now getting a bunch of 

    xmonad: stack: runProcess: invalid argument (Bad file descriptor)

which is especially weird since both 

    stack build --silent

and 

    stack ghc --stack-yaml stack.yaml -- --make xmonad.hs -i -ilib -fforce-recomp -main-is main -v0

work just fine when entered from the console.  I'll try to debug this later today

Reply:
Ah I think the reason is that [runProcess](https://hackage.haskell.org/package/process-1.6.13.2/docs/src/System.Process.html#runProcess) closes the handles that it gets: 

> Any Handles passed to runProcess are placed immediately in the closed state.

Reply:
Oh, sorry, I didn't bother running the code, and of course it bit me. We need linear types! :-)

> Any Handles passed to runProcess are placed immediately in the closed state.

I'll fix this.

Reply:
79682c5bb144ba7b872a07823feb5618e52d301b works for me

Reply:
Works :+1:  

The error messages this generates when something goes bad are quite funny 

``` console
Errors detected while compiling xmonad config: /home/slot/.config/xmonad/xmonad.hs
ExitFailure 1

Please check the file for errors.
```

but I'm not sure we can do anything about that (this is just a byproduct of the `--silent` flag).  

All in all this looks good to me now (and I also can't reproduce the problem mentioned below anymore)

Reply:
It might be quite helpful to print the executed commands into the error file to make its output less cryptic. I'll add that and merge this now to get as much testing as possible.